File: app/src/main/java/com/itachi1706/busarrivalsg/Services/BusStorage.java
package com.itachi1706.busarrivalsg.Services;

import android.content.SharedPreferences;
import android.util.Log;

import com.google.gson.Gson;
import com.itachi1706.busarrivalsg.gsonObjects.offline.BusArrayJSON;
import com.itachi1706.busarrivalsg.gsonObjects.offline.BusJSON;
import com.itachi1706.busarrivalsg.objects.BusServices;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Objects;

/**
 * Created by Kenneth on 17/6/2015
 * for SingBuses in package com.itachi1706.busarrivalsg
 */
public class BusStorage {

    private static final String TAG = "BusStorage";

    public static void addNewBus(BusServices bus, SharedPreferences prefs){
        String serviceNo = bus.getServiceNo();
        String operator = bus.getOperator();
        String stopID = bus.getStopID();
        String stopName = bus.getStopName();

        JSONArray arr = getExistingJSONString(prefs);
        JSONObject obj = new JSONObject();
        JSONObject main = new JSONObject();
        try {
            obj.put("service", serviceNo);
            obj.put("operator", operator);
            obj.put("stop", stopID);
            obj.put("stopName", stopName);
            arr.put(obj);
            main.put("storage",arr);
            prefs.edit().putString("stored", main.toString()).apply();
        } catch (JSONException e) {
            Log.e(TAG, "Error adding new bus to storage");
            Log.e(TAG, "Error message: " + e.getLocalizedMessage());
        }
    }

    private static JSONArray getExistingJSONString(SharedPreferences pref) {
        String json = pref.getString("stored", null);
        if (json == null) {
            return new JSONArray();
        }
        try {
            JSONObject obj = new JSONObject(json);
            return obj.getJSONArray("storage");
        } catch (JSONException e) {
            Log.e(TAG, "Error getting existing JSON String");
            Log.e(TAG, "Error message: " + e.getLocalizedMessage());
        }
        return new JSONArray();
    }

    public static ArrayList<BusServices> getStoredBuses(SharedPreferences pref){
        String json = pref.getString("stored", "-");
        if (json.equals("-")){
            return null;
        }

        ArrayList<BusServices> services = new ArrayList<>();

        Gson gson = new Gson();
        BusArrayJSON busArray = gson.fromJson(json, BusArrayJSON.class);
        for (BusJSON b : Objects.requireNonNull(busArray.getStorage())){
            BusServices bs = new BusServices();
            bs.setOperator(b.getOperator());
            bs.setServiceNo(b.getService());
            bs.setStopID(b.getStop());
            if (b.getStopName() != null)
                bs.setStopName(b.getStopName());
            bs.setObtainedNextData(false);
            services.add(bs);
        }

        return services;
    }

    public static void updateBusJSON(SharedPreferences pref, ArrayList<BusServices> newServices){
        pref.edit().remove("stored").apply();
        if (newServices.size() != 0) {
            for (BusServices s : newServices) {
                addNewBus(s, pref);
            }
        }
    }

    public static boolean hasFavourites(SharedPreferences pref){
        String check = pref.getString("stored", "wot");
        return !check.equals("wot");
    }
}

File: app/src/main/java/com/itachi1706/busarrivalsg/services/BusStorage.kt
package com.itachi1706.busarrivalsg.services

import android.content.SharedPreferences
import androidx.core.content.edit
import com.itachi1706.busarrivalsg.objects.BusServices
import com.itachi1706.busarrivalsg.objects.json.offline.BusArrayJSON
import com.itachi1706.helperlib.helpers.LogHelper
import kotlinx.serialization.json.Json
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class BusStorage(val prefs: SharedPreferences) {

    fun addNewBus(bus: BusServices) {
        val arr = getExistingJsonString()
        val obj = JSONObject()
        val main = JSONObject()

        try {
            obj.put("service", bus.serviceNo)
            obj.put("operator", bus.operator)
            obj.put("stop", bus.stopID)
            obj.put("stopName", bus.stopName)
            arr.put(obj)
            main.put("storage", arr)
            prefs.edit { putString(PREF_STORED_NAME, main.toString()) }
        } catch (e: JSONException) {
            LogHelper.e(TAG, "Error adding new bus to storage")
            LogHelper.e(TAG, "Error message: ${e.localizedMessage}")
        }
    }

    fun getStoredBuses(): List<BusServices> {
        val json = prefs.getString(PREF_STORED_NAME, null)
        if (json.isNullOrEmpty()) {
            return emptyList()
        }

        val services = ArrayList<BusServices>()

        val jsonConfig = Json { ignoreUnknownKeys = true }
        val busArray = jsonConfig.decodeFromString<BusArrayJSON>(json)
        if (busArray == null || busArray.storage.isNullOrEmpty()) {
            return emptyList()
        }
        for (bus in busArray.storage) {
            val service = BusServices(
                serviceNo = bus.service ?: "",
                operator = bus.operator ?: "",
                stopID = bus.stop ?: "",
                stopName = bus.stopName ?: "",
                isObtainedNextData = false
            )
            services.add(service)
        }

        return services
    }

    fun updateBusJson(newServices: List<BusServices>) {
        prefs.edit{ remove(PREF_STORED_NAME) }
        if (newServices.isNotEmpty()) {
            newServices.forEach { addNewBus(it) }
        }
    }

    fun hasFavourites(): Boolean {
        val check = prefs.getString(PREF_STORED_NAME, "wot")
        return check != "wot"
    }

    private fun getExistingJsonString(): JSONArray {
        val json = prefs.getString(PREF_STORED_NAME, null)
        if (json.isNullOrEmpty()) {
            return JSONArray()
        }

        return try {
            val obj = JSONObject(json)
            obj.getJSONArray("storage")
        } catch (e: Exception) {
            LogHelper.e(TAG, "Error getting existing JSON string")
            LogHelper.e(TAG, "Error message: ${e.localizedMessage}")
            JSONArray()
        }
    }

    companion object {
        const val TAG = "BusStorage"
        private const val PREF_STORED_NAME = "stored"
    }
}